torch_geometric/nn/summary.py

```diff
@@ -0,0 +1,156 @@
+
+import torch
+import torch.nn as nn
+
+
+def summary(
+    model: nn.Module,
+    *inputs: Sequence[Any],
+    device: torch.device = 'cpu',
+    max_depth: int = 3,
+) -> str:
+    r"""Summarizes the given PyTorch model. Summarized information includes
+    (1) Layer names, (2) Input/output shapes,
+    (3) # of parameters, and (4) Excutation time.
+
+    .. code-block:: python
+
+        import torch
+        from torch_geometric.nn import summary
+        from torch_geometric.nn.models import GCN
+
+        model = GCN(128, 64, 2, out_channels=32)
+        x = torch.randn(100, 128)
+        edge_index = torch.randint(100, size=(2, 20))
+        print(summary(model, x, edge_index))
+
+    Args:
+        model (nn.Module): PyTorch model to summarize.
+        inputs (Sequence[Any]): Arguments for the model's
+            forward pass.
+        device (torch.device): Device for model and input_data.
+            (default: :obj:`"cpu"`)
+        max_depth (int): Depth of nested layers to display.
+            Nested layers below this depth will not be displayed
+            in the summary. (default: :obj:`"3"`)
+    """
+    def register_pre_hook(info):
+        def hook(module, input):
+            info['input_time'].append(perf_counter())
+
+        return hook
+
+    def register_hook(info):
+        def hook(module, input, output):
+            info['input_shape'].append(get_shape(input))
+            info['output_shape'].append(get_shape(output))
+            info['output_time'].append(perf_counter())
+
+        return hook
+
+    hooks = {}
+    depth = 0
+    stack = [(get_name(model), model, depth)]
+
+    info_list = []
+    input_shape = defaultdict(list)
+    output_shape = defaultdict(list)
+    intput_time = defaultdict(list)
+    output_time = defaultdict(list)
+    while stack:
+        var_name, module, depth = stack.pop()
+        module_id = id(module)
+        if module_id in hooks:
+            for hook in hooks[module_id]:
+                hook.remove()
+
+        info = {}
+        info['name'] = var_name
+        info['layer'] = module
+        info['input_shape'] = input_shape[id(module)]
+        info['output_shape'] = output_shape[id(module)]
+        info['input_time'] = intput_time[id(module)]
+        info['output_time'] = output_time[id(module)]
+        info['depth'] = depth
+        para = sum(p.numel() for p in module.parameters())
+        info['#param'] = f"{para:,}" if para > 0 else "--"
+        info_list.append(info)
+        hooks[module_id] = (
+            module.register_forward_pre_hook(register_pre_hook(info)),
+            module.register_forward_hook(register_hook(info)),
+        )
+        module_items = reversed(module._modules.items())
+        stack += [(f"({name}){get_name(mod)}", mod, depth + 1)
+                  for name, mod in module_items if mod is not None]
+
+    # set `eval` mode
+    model.eval()
+    # make a forward pass
+    device = torch.device(device)
+    with torch.no_grad():
```

Comment:
Let's drop this here :)

Reply:
Sorry, did you mean to drop this line `with torch.no_grad()`? Any reason?

Reply:
No I meant the `device` :D